enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -4790,9 +4799,103 @@ class AdjointGenerator
     assert(called);
     assert(gutils->getWidth() == 1);
 
-    IRBuilder<> BuilderZ(gutils->getNewFromOriginal(&call));
+    auto newCall = gutils->getNewFromOriginal(&call);
+    IRBuilder<> BuilderZ(newCall);
     BuilderZ.setFastMathFlags(getFast());
 
+    enum class MPI_Op_Kind { SUM, MIN, MAX, MINLOC, MAXLOC, UNKNOWN };
```

Comment:
I've added a few helper functions and maps at the start of handleMPI that I extracted from MPI_Allreduce handling, and extended.